mautic/mautic
# [UX] Toggles

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🔴 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 



## Description

Switches are going to replace the Yes/No buttons, keeping the same backend way to handle data saving to ensure compatibility with absolutely all existing functionalities and JS code.

![image](https://github.com/user-attachments/assets/4a011065-0d91-4fe1-a110-9f48dddb9537)

This PR also features a mini toggle where space is an issue, when you're adding them to tables or Reports filter rows.

![image](https://github.com/user-attachments/assets/055d047b-dbd3-48f4-ad0f-19f058c3cc09)

Have a look at this example **(not a real implementation, just made it look small to show)**. The "Use theme style" field is using a small toggle when compared to the others.

![image](https://github.com/user-attachments/assets/74dd240c-0f1a-4cfd-954a-1bdc92e350a2)

The beauty here is the UX improvement, as it allows to have custom labels according to the field status. THey help to understand the consequences of each state. Look at the status labels for the Kiosk mode:

| Disabled                                 | Enabled
| -------------------------------------- | ---
| ![image](https://github.com/user-attachments/assets/2a3e6a3e-b4a7-4eb9-9c1e-2d7de47354c9)                                      | ![image](https://github.com/user-attachments/assets/92c563e2-3bbd-4cd3-a066-1b8af211d5f3)


In many cases, they can replace tooltips.
These custom labels are set using the existing PHP functionality. We have a class for the Yes/No buttons that provides a default value for no_label and yes_label (below):

```
class YesNoButtonGroupType extends AbstractType
{
    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults(
            [
                'choices'           => fn (Options $options): array => [
                    $options['no_label']  => $options['no_value'],
                    $options['yes_label'] => $options['yes_value'],
                ],
                'choice_value'      => function ($choiceKey) {
                    if (null === $choiceKey || '' === $choiceKey) {
                        return null;
                    }

                    return (is_string($choiceKey) && !is_numeric($choiceKey)) ? $choiceKey : (int) $choiceKey;
                },
                'expanded'          => true,
                'multiple'          => false,
                'label_attr'        => ['class' => 'control-label'],
                'label'             => 'mautic.core.form.active',
                'placeholder'       => false,
                'required'          => false,
                'no_label'          => 'mautic.core.form.no',     // labels here
                'no_value'          => 0,
                'yes_label'         => 'mautic.core.form.yes',
                'yes_value'         => 1,
            ]
        );
    }
}
```

The custom labels can be set when adding fields to the builder, providing a new translation string for each:

```
        $builder->add('inKioskMode', YesNoButtonGroupType::class, [
            'label'     => 'mautic.form.form.kioskmode',
            'attr'      => [
                'tooltip' => 'mautic.form.form.kioskmode.tooltip',
            ],
            'yes_label' => 'mautic.form.form.kioskmode.yes',
            'no_label'  => 'mautic.form.form.kioskmode.no',
        ]);
```






---
### 📋 Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Go to Forms > New > Toggle the option Kiosk mode
3. Go to Segments > New > Toggle Visible for other users > Check the label for No (Only me)
4. Go to Points > Point Actions > New > Toggle Is repeatable? and see two different labels
5. Go to Reports > New > Toggle Visible for all logged-in users to see the No label
6. Go to Custom Fields > New > Toggle the Indexable field
7. Go to Custom fields > New > Check a new and improved tooltip for Unique identifier
